### ISSUE TITLE
Illogical error: selenium-java:3.0.1 requiring selenium-java:2.53.0

### ISSUE THREAD

**Issue by kriegaex** [mojohaus/extra-enforcer-rules#28]
Hi. I have managed a dependency version like this:

```
  <properties>
    <version.selenium>3.0.1</version.selenium>
  </properties>

      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-java</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
```

Now look at this error with extra enforcer rules 1.0-beta-6:

```
[INFO] --- maven-enforcer-plugin:1.4.1:enforce (enforcer-rules-warning) @ diamond-square ---
[WARNING] 
Dependency convergence error for org.seleniumhq.selenium:selenium-java:3.0.1 paths to dependency are:
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
and
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
    +-org.seleniumhq.selenium:selenium-java:2.53.0

[WARNING] Rule 1: org.apache.maven.plugins.enforcer.DependencyConvergence failed with message:
Failed while enforcing releasability the error(s) are [
Dependency convergence error for org.seleniumhq.selenium:selenium-java:3.0.1 paths to dependency are:
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
and
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
    +-org.seleniumhq.selenium:selenium-java:2.53.0
]
```

Especially this:
```
  +-org.seleniumhq.selenium:selenium-java:3.0.1
    +-org.seleniumhq.selenium:selenium-java:2.53.0
```

This seems to be a bug. Please tell me if you need more info.

**Comment by khmarbaise** [comment]
Can you post your pom files cause it could also be the case that you defined the selenium java in an other pom which might be used as transitive dependency and/or in a different scope? Apart from that can you check with current version of extra-enforcer-rules...

**Comment by kriegaex** [comment]
It is actually a single POM file. I have taken some time to strip it down to a near-minimum size for you. Just call `mvn compile` and see what happens.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>de.scrum-master</groupId>
  <artifactId>extra-enforcer-rules-problem</artifactId>
  <version>1.0-SNAPSHOT</version>

  <prerequisites>
    <maven>3.1.0</maven>
  </prerequisites>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    <version.selenium>3.0.1</version.selenium>
  </properties>

  <build>

    <pluginManagement>
      <plugins>

        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-compiler-plugin</artifactId>
          <version>3.6.0</version>
        </plugin>

        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-enforcer-plugin</artifactId>
          <version>1.4.1</version>
          <executions>
            <execution>
              <id>enforcer-rules-error</id>
              <goals>
                <goal>enforce</goal>
              </goals>
              <configuration>
                <fail>true</fail>
                <skip>false</skip>
                <rules>
                  <requireReleaseDeps>
                    <onlyWhenRelease>true</onlyWhenRelease>
                  </requireReleaseDeps>
                </rules>
              </configuration>
            </execution>
            <execution>
              <id>enforcer-rules-warning</id>
              <goals>
                <goal>enforce</goal>
              </goals>
              <configuration>
                <fail>false</fail>
                <skip>false</skip>
                <rules>
                  <dependencyConvergence/>
                </rules>
              </configuration>
            </execution>
          </executions>
          <dependencies>
            <dependency>
              <groupId>org.codehaus.mojo</groupId>
              <artifactId>extra-enforcer-rules</artifactId>
              <version>1.0-beta-6</version>
            </dependency>
          </dependencies>
        </plugin>

      </plugins>
    </pluginManagement>

    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-enforcer-plugin</artifactId>
      </plugin>
    </plugins>
  </build>

  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-java</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-support</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>com.codeborne</groupId>
        <artifactId>phantomjsdriver</artifactId>
        <version>1.3.0</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-firefox-driver</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-chrome-driver</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-ie-driver</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-edge-driver</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-opera-driver</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>selenium-remote-driver</artifactId>
        <version>${version.selenium}</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.apache.httpcomponents</groupId>
        <artifactId>httpclient</artifactId>
        <version>4.5.2</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>commons-codec</groupId>
        <artifactId>commons-codec</artifactId>
        <version>1.10</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>commons-io</groupId>
        <artifactId>commons-io</artifactId>
        <version>2.5</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>org.eclipse.jetty.websocket</groupId>
        <artifactId>websocket-client</artifactId>
        <version>9.2.18.v20160721</version>
      </dependency>
      <dependency>
        <groupId>org.seleniumhq.selenium</groupId>
        <artifactId>htmlunit-driver</artifactId>
        <version>2.23.2</version>
        <scope>test</scope>
      </dependency>
      <dependency>
        <groupId>io.github.bonigarcia</groupId>
        <artifactId>webdrivermanager</artifactId>
        <version>1.5.1</version>
        <scope>test</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <dependency>
      <groupId>org.seleniumhq.selenium</groupId>
      <artifactId>selenium-java</artifactId>
      <scope>test</scope>
    </dependency>
  </dependencies>

</project>
```

**Comment by kriegaex** [comment]
Any news here? BTW, when manually inspecting the dependencies, I saw that instead of 

```
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
and
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
    +-org.seleniumhq.selenium:selenium-java:2.53.0
```

it should be

```
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-org.seleniumhq.selenium:selenium-java:3.0.1
and
+-de.scrum-master.tdd:diamond-square:1.0-SNAPSHOT
  +-com.codeborne:phantomjsdriver:1.3.0
    +-org.seleniumhq.selenium:selenium-java:2.53.0
```

because `phantomjsdriver` is what really declares the `selenium-java:2.53.0` dependency. It is an obvious bug and I would like to get it fixed or told how I can fix it temporarily in a patched snapshot. Nothing is worse than false or incomprehensible warnings when trying to keep dependency management clean or spot possible problems.

**Comment by kriegaex** [comment]
Why was this issue closed? I provided a reproducer and got no feedback or explanation whatsoever for almost five years.

**Comment by slachiewicz** [comment]
Is this still valid after 5years?

**Comment by kriegaex** [comment]
Sorry, do you always shoot first and ask questions later? I also find the approach questionable, to let an issue with a reproducer sit there and rot for 5 years and then just close it, as if doing nothing would automatically invalidate it. Or was it rather "let's just close it and see if we get lucky and nobody protests"?

How about reproducing the problem first as a maintainer and find the root cause? Or if by chance upgrading a version of Enforcer or Extra Enforcer Rules makes the issue go away, find the commit which fixed it - `git bisect` could be a helpful tool there - and link to it in an explanatory comment? Are you expecting me to do that on your behalf now, kind of as a reward for having been ignored for years?

**Comment by kriegaex** [comment]
FWIW, the issue is still reproducible with Extra Enforcer Rules 1.4 and up to Enforcer 3.0.0-M3. When using Enforcer 3.0.0 with either Extra Enforcer Rules 1.0-beta-6 or 1.4, there is nothing reported at all by same rule and the build passes. Should there not be an error report, but with the correct listing of problematic dependencies? I am not sure what ought to happen there, but silence also does not seem to be OK. Probably one of the project's committers can enlighten me here.

**Comment by kriegaex** [comment]
> FWIW, the issue is still reproducible with Extra Enforcer Rules 1.4

Well, as I just realised, `dependencyConvergence` is a built-in Enforcer rule rather than an extra rule, so it is not really a surprise that changing the extra rules version does not change the behaviour but that changing the enforcer version as such does. Maybe some 5 years ago some maintainer could have told me that this issue belongs to the plugin project, not to the extra rules project, and that filing it here was an oversight on my part. But obviously, nobody even tried to analyse my reproducer POM. I really do not know what more can be expected of an issue reporter than to provide a simple reproducer project, which I did upon request, but nobody cared enough to even take a look. Otherwise I would have moved this report into the plugin project, where it might have been resolved years ago instead of only between 3.0.0-M3 and the final release, i.e. some time between November 2019 and July 2021.

**Comment by kriegaex** [comment]
OK, I added an integration test with a simplified version of my POM above to a local clone of Maven Enforcer and bisected it. The commit which fixed it on 2021-07-13 is https://github.com/apache/maven-enforcer/commit/ca40308fd58c45e638a35768b3966b5680d4c60e, as follows:

```bash
# Set up reverse bisecting to find the commit which fixed the problem
git bisect start --term-new fixed --term-old broken
git bisect broken enforcer-3.0.0-M3
git bisect fixed enforcer-3.0.0

# Run prepared IT and afterwards specify if the behaviour was 'fixed' or 'broken'.
# Do this in a loop until we found the commit which fixed the problem
run mvn -P run-its -Denforcer.skip=true -Drat.skip=true -DskipTests -Dinvoker.test=dependency-convergence-problem verify
git bisect fixed|broken

# Finally, display the bisection log
git bisect log
```

```text
$ git bisect log
git bisect start '--term-new' 'fixed' '--term-old' 'broken'
# broken: [edc174b9ce8fc45974343d075a75fcb23981bd91] [maven-release-plugin] prepare release enforcer-3.0.0-M3
git bisect broken edc174b9ce8fc45974343d075a75fcb23981bd91
# fixed: [b1b22822174bc92857a2e674c9a024035ee6d7cd] [maven-release-plugin] prepare release enforcer-3.0.0
git bisect fixed b1b22822174bc92857a2e674c9a024035ee6d7cd
# broken: [355f8452e7367d1f9b1095cd7d6851be349fecb0] [MENFORCER-350] Added few tests to validate issue
git bisect broken 355f8452e7367d1f9b1095cd7d6851be349fecb0
# broken: [eadd63210b0d640b0b2b2672d4499a1771ff4052] [MENFORCER-257] RequireActiveProfile should respect inherited activated profiles
git bisect broken eadd63210b0d640b0b2b2672d4499a1771ff4052
# fixed: [0ffd8d92b9ca4ce09eb414ce8ef8f111fb1484cf] Bump jsr305 from 3.0.0 to 3.0.2
git bisect fixed 0ffd8d92b9ca4ce09eb414ce8ef8f111fb1484cf
# fixed: [cecb4b330de62be0f7033733c839322b9d4cfe95] Fix projectname for extension
git bisect fixed cecb4b330de62be0f7033733c839322b9d4cfe95
# broken: [a6949503a48702d59dabdf68cc37a87ed17a8b21] [MENFORCER-329] Example for writing a custom rule should be upgraded
git bisect broken a6949503a48702d59dabdf68cc37a87ed17a8b21
# fixed: [ca40308fd58c45e638a35768b3966b5680d4c60e] [MENFORCER-277] Upgrade maven-dependency-tree to 3.x
git bisect fixed ca40308fd58c45e638a35768b3966b5680d4c60e
# broken: [fbc505a3df144464d9210839409e856af27fcb50] [MENFORCER-387] Require Java 8
git bisect broken fbc505a3df144464d9210839409e856af27fcb50
# first fixed commit: [ca40308fd58c45e638a35768b3966b5680d4c60e] [MENFORCER-277] Upgrade maven-dependency-tree to 3.x
```

So it looks like upgrading `maven-dependency-tree` from 2.2. to 3.1.0 and adjusting the corresponding API calls has indirectly taken care of removing this issue. As you can see, the commit is a quite recent one. This problem should have been taken care of years ago, as due to me raising the issue here, it was a known one.

@slachiewicz, because you also contribute to Maven Enforcer Plugin, I would like to ask you if you thisnk it might be a good idea for me to create a belated issue and corresponding PR with an IT for that project in order to catch possible future regressions.

**Comment by slachiewicz** [comment]
Thank you for your time to find root cause. Good that this is now fixed. Not always we as commiters (I'm not maintainer) have enough knowledge or more important - time to go deeply to each issue reported. So to minimize backlog I'm closing issues that looks not so hot to focus on something newly reported. Usually, updating libs gives chance that other project fixed issue for us. I  don't want to enforce someone to reproduce issue reported 5 years ago.
Good that you gave knowledge, please feel free to create additional integration tests in enforcer, help is always welcomed. Please check also other Maven issues - it's now about 1900 reported across 96 repos plus many more at mohohous and plexus.

Sorry for inconvenience
